automerge/src/transaction.rs

```diff
@@ -0,0 +1,203 @@
+            .unwrap()
+            .insert(self.doc, obj, index, value)
+    }
+
+    pub fn inc<P: Into<Prop>>(
+        &mut self,
+        obj: &ExId,
+        prop: P,
+        value: i64,
+    ) -> Result<(), AutomergeError> {
+        self.inner.as_mut().unwrap().inc(self.doc, obj, prop, value)
+    }
+
+    pub fn del<P: Into<Prop>>(&mut self, obj: &ExId, prop: P) -> Result<(), AutomergeError> {
+        self.inner.as_mut().unwrap().del(self.doc, obj, prop)
+    }
+
+    /// Splice new elements into the given sequence. Returns a vector of the OpIds used to insert
+    /// the new elements
+    pub fn splice(
+        &mut self,
+        obj: &ExId,
+        pos: usize,
+        del: usize,
+        vals: Vec<Value>,
+    ) -> Result<Vec<ExId>, AutomergeError> {
+        self.inner
+            .as_mut()
+            .unwrap()
+            .splice(self.doc, obj, pos, del, vals)
+    }
+
+    pub fn splice_text(
+        &mut self,
+        obj: &ExId,
+        pos: usize,
+        del: usize,
+        text: &str,
+    ) -> Result<Vec<ExId>, AutomergeError> {
+        self.inner
+            .as_mut()
+            .unwrap()
+            .splice_text(self.doc, obj, pos, del, text)
+    }
+
+    pub fn keys(&self, obj: &ExId) -> Vec<String> {
+        self.doc.keys(obj)
+    }
+
+    pub fn keys_at(&self, obj: &ExId, heads: &[ChangeHash]) -> Vec<String> {
+        self.doc.keys_at(obj, heads)
+    }
+
+    pub fn length(&self, obj: &ExId) -> usize {
+        self.doc.length(obj)
+    }
+
+    pub fn length_at(&self, obj: &ExId, heads: &[ChangeHash]) -> usize {
+        self.doc.length_at(obj, heads)
+    }
+
+    pub fn text(&self, obj: &ExId) -> Result<String, AutomergeError> {
+        self.doc.text(obj)
+    }
+
+    pub fn text_at(&self, obj: &ExId, heads: &[ChangeHash]) -> Result<String, AutomergeError> {
+        self.doc.text_at(obj, heads)
+    }
+
+    pub fn value<P: Into<Prop>>(
+        &self,
+        obj: &ExId,
+        prop: P,
+    ) -> Result<Option<(Value, ExId)>, AutomergeError> {
+        self.doc.value(obj, prop)
+    }
+
+    pub fn value_at<P: Into<Prop>>(
+        &self,
+        obj: &ExId,
+        prop: P,
+        heads: &[ChangeHash],
+    ) -> Result<Option<(Value, ExId)>, AutomergeError> {
+        self.doc.value_at(obj, prop, heads)
+    }
+
+    pub fn values<P: Into<Prop>>(
+        &self,
+        obj: &ExId,
+        prop: P,
+    ) -> Result<Vec<(Value, ExId)>, AutomergeError> {
+        self.doc.values(obj, prop)
+    }
+
+    pub fn values_at<P: Into<Prop>>(
+        &self,
+        obj: &ExId,
+        prop: P,
+        heads: &[ChangeHash],
+    ) -> Result<Vec<(Value, ExId)>, AutomergeError> {
+        self.doc.values_at(obj, prop, heads)
+    }
+}
+
+// If a transaction is not commited or rolled back manually then it can leave the document in an
+// intermediate state.
+// This defaults to rolling back the transaction to be compatible with `?` error returning before
+// reaching a call to `commit`.
```

Comment:
This means we could be doing quite a bit of work in the Drop implementation so we should document that `Drop` behaviour on the `Transaction` I think.

Reply:
Added.